cassiopeia/solver/GreedySolver.py

```diff
@@ -139,40 +135,43 @@ def _solve(samples):
             return root
 
         self.tree = nx.DiGraph()
-        for i in range(len(self.node_mapping)):
+        samples = list(self.unique_character_matrix.index)
+        for i in samples:
             self.tree.add_node(i)
-        _solve(list(range(self.unique_character_matrix.shape[0])))
+        _solve(samples)
         # Collapse 0-mutation edges and append duplicate samples
         self.tree = solver_utilities.collapse_tree(
-            self.tree, True, self.unique_character_matrix, self.missing_char
+            self.tree, True, self.character_matrix, self.missing_char
         )
-        self.tree = nx.relabel_nodes(self.tree, self.node_mapping)
         self.tree = self.add_duplicates_to_tree(self.tree)
         return self.tree
 
     def compute_mutation_frequencies(
-        self, samples: List[int]
+        self, samples: List[str]
     ) -> Dict[int, Dict[int, int]]:
         """Computes the number of samples in a character matrix that have each
         character/state mutation.
 
-        Generates a dictionary that maps each character to a dictionary of state/
-        sample frequency pairs, allowing quick lookup. Subsets the character matrix
-        to only include the samples in the sample set.
+        Computes the frequency of each mutation in the character data of the
+        sample set. Duplicate cells have their frequencies included. Generates
+        a dictionary that maps each character to a dictionary of state/sample
+        frequency pairs, allowing quick lookup.
 
         Args:
-            cm: The character matrix from which to calculate frequencies
-            missing_char: The character representing missing values
-            samples: The set of relevant samples in calculating frequencies
+            samples: The set of relevant samples in calculating frequencies,
+                represented by their names in the original character matrix
 
         Returns:
-            A dictionary containing frequency information for each character/state
-            pair
+            A dictionary containing frequency information for each character/
+            state pair
 
         """
-        subset_cm = self.unique_character_matrix[samples, :]
+        sample_indices = solver_utilities.convert_sample_names_to_indices(
+            self.unique_character_matrix.index, samples
+        )
+        subset_cm = self.unique_character_matrix.to_numpy()[sample_indices, :]
```

Comment:
Just to confirm - the reason you are doing this is because it is a lot faster to to get entries from a numpy matrix with integer indices than to access a pandas dataframe with `.loc`, correct? 

Are we sure that the new function `convert_sample_names_to_indices` doesn't add the same overhead? Was there a reason why you went from having a `node_mapping` helper variable in the class?

Reply:
I should have added -- I see later on how you rely on this `convert_sample_names_to_indices` function in contexts outside of a class, so I actually think it makes sense in general. Just making sure that we aren't creating extra lines of code here if it's the same overhead as `subset_cm = self.unique_character_matrix.lo[sample_indices].to_numpy()`

Reply:
I did some runtime testing by using a jupyter notebook, and to my understanding the mapping to indices actually takes about as long as _one_ iloc or loc call. So I think it would indeed save us a lot of time, especially in the graph methods where iloc or loc would be called many many times. I saw this in benchmarking where I got order of magnitudes speedup

Reply:
Fantastic - thanks for the clarification here. So indeed, `dataframe.to_numpy()[indices,:]` is a lot faster than `dataframe.loc[samples].to_numpy`. Interesting!

Reply:
As far as the removing of the node_mapping attribute, I thought that keeping these attributes around might be confusing to a user who is looking at the class attributes, given that the class level is supposed to respect the string interface. I thought it might be nice just to have individual method handle its own conversion on the backend. If you feel that this isn't an issue and we could save some overhead, then I can change it back, but overall I think the runtime contribution of the conversions should be negligible.

Reply:
I actually think it's a nice utility to have around and definitely cuts down on duplicate code. Good call!

Reply:
Actually sorry to clarify, in this instance it's about the same of magnitude, so I will reduce the lines of code here.